### ISSUE TITLE
generating allOf with discriminator

### ISSUE THREAD

**Issue by felschr** [domaindrivendev/Swashbuckle.AspNetCore#1973]
According to the OpenAPI specification:
> OpenAPI lets you combine and extend model definitions using the `allOf` keyword. `allOf` takes an array of object definitions that are used for independent validation but together compose a single object. Still, it **does not imply a hierarchy between the models. For that purpose, you should include the `discriminator`**.

— https://swagger.io/docs/specification/data-models/oneof-anyof-allof-not/#allof

There is an example of `allOf` being used with `discriminator` where it's being set on the base class `Pet`:
https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md#discriminatorObject
This is different from using a `discriminator` with `oneOf`.

In fact [openapi-generator](https://github.com/OpenAPITools/openapi-generator) has [dropped support for `allOf` inheritance without a `discriminator`](https://github.com/OpenAPITools/openapi-generator/issues/8495#issuecomment-766647017). It's stopped generating any inheritance information (e.g. no `extends` in typescript) if the `discriminator` is missing.

As far as I can see Swashbuckle only adds discriminator information when using `oneOf` for inheritance.
My config uses `options.UseAllOfForInheritance();` and I've added `SwaggerSubType` annotations. I was previously using
```
options.EnableAnnotations(
    enableAnnotationsForInheritance: true,
    enableAnnotationsForPolymorphism: true);
```
to generate `oneOf` & `discriminator` information but they're interpreted as unions (at least by openapi-generator) which might not be wanted. Additionally openapi-generator currently has some issues with union generation which prevents us from using `oneOf` anyway.

**TL;DR**
My question is whether `allOf` can already be used with `discriminator` and I just haven't found the option yet or whether it's currently not supported. More specifically does this work when using `SwaggerSubType` annotations?

**Comment by felschr** [comment]
I think this should probably work like this:
If `options.UseOneOfForPolymorphism()` and `options.EnableAnnotations()` with `enableAnnotationsForInheritance: true` is used it should generate the `oneOf` with `discriminator` set on the same schema.

And if `enableAnnotationsForInheritance` is instead used with `options.UseAllOfForPolymorphism()` it should only use `allOf` and set the `discriminator` on the base class. No `oneOf` code should be generated in this case.

**Comment by domaindrivendev** [comment]
The code in this area is getting extremely complex and as it happens, I'm trying to get a small refactor done for the next major release (`6.0.0`), which I'm striving to get out ASAP (weeks not months), so I'll try look into this.

I've been basing the SB implementation of Inheritance & Polymorphism on [these docs](https://swagger.io/docs/specification/data-models/inheritance-and-polymorphism/), where it seems the `discriminator` metadata is only applicable on a schema with `oneOf`. In those docs, there's no mention of the alternative approach of applying it in the parent schema definition to avoid redundancy. In other words, the following statement from the docs you referenced, is news to me:

> To avoid redundancy, the discriminator MAY be added to a parent schema definition, and all schemas comprising the parent schema in an allOf construct may be used as an alternate schema

With this new info, I now wonder if it may help simplify the SB implementation internally. In any case, I'll try find the time to dig deeper in the next few days, and possibly get support baked into the `6.0.0` release - not promising anything, but stay tuned for an update 😉

**Comment by domaindrivendev** [comment]
Supporting the various combinations of `UseAllOfForInheritance` and `UseOneOfForPolymorphism` is proving to be extremely complex and unworkable with the current design. With that said, I don't believe supporting `oneOf` _without_ a corresponding `allOf` reference to the base type schema on each possible schema is particularly useful. So, I'm leaning toward a solution where `UseOneOfForPolymorphism` automatically _implies_ `UseAllOfForInheritance`, and the `discriminator` metadata is _always_ applied to the base type schema definition.

@felschr I have this implemented in current master, and would be grateful if you could pull down the latest preview nuget from [myget.org](https://myget.org/feed/domaindrivendev/package/nuget/Swashbuckle.AspNetCore) to see if it solves your issue.

**Comment by felschr** [comment]
Thank you for looking into this so quickly!

This is probably not related to the inheritance changes but I'm getting the following error from `openapi-generator` after updating Swashbuckle from `5.6.0` to the NuGet preview:
```
        -attribute paths.'/calendar/api/users/{userId}/calendar/recommendations'(post).requestBody.content.'application/json-patch+json'.schema.default is not of type `array`
        -attribute paths.'/calendar/api/users/{userId}/calendar/recommendations'(post).parameters.[start].schemas.default=`null` is not of type `date-time`
```
I can see that previously the schema didn't have any `default` values at all for those types.
It might have to do with my custom schema filter for nullable attributes.

Also, I just noticed that the latest MyGet preview (6.0.0-preview-1528) was published on 2020-01-19.

**Comment by domaindrivendev** [comment]
Hi @felschr - it seems there was some issue with my CI/CD that was preventing latest builds going to myget.org.

I've fixed this now and pushed `6.0.0-preview-1549`, and then tested out the new functionality and it appears to work well. Could you give it another try when you get a chance?

**Comment by felschr** [comment]
Alright, I gave it a try and the default value issues are gone now and `discriminator`s are being generated on the base classes 🥳.
But it's still generating `oneOf` where I'd just like to have the base class.

That's with my config:
```cs
options.UseAllOfForInheritance();
options.EnableAnnotations(
    enableAnnotationsForInheritance: true,
    enableAnnotationsForPolymorphism: true);
```

Do you think it would be possible to support this case as well?

My problem with `oneOf` for inheritance is that `openapi-generator` will create unions for it.
I think having unions to represent inheritance is generally undesirable.

**Comment by felschr** [comment]
Oh, never mind, I think setting `enableAnnotationsForPolymorphism: false` does what I want. I'm gonna test this a bit.

**Comment by felschr** [comment]
I haven't tested using this with `UseOneOfForPolymorphism()` but with `UseAllOfForInheritance()` it does exactly what we needed.
Even `openapi-generator` is happy with this. Great work 🥳!
